Code folding UI & code style tweaks

peterflynn opened the pull request:
~~We should land PR #10850 first, and then the diff for this will be simplified.~~ _(done)_

~~For now, this link previews what the diff will be after 10850 lands: https://github.com/thehogfather/brackets/compare/issue10846...adobe:pflynn/codefolding-tweaks~~

Improve visual appearance of code folding:
- Fainter, smaller gutter triangle icons
- Gutter triangle icons become less faint on mouseover (or when collapsed)
- Vertically align gutter triangle icons with text better
- Change inline collapsed placeholder to more readable arrow icon
- Better colors in dark theme
- Less horizontal gap between line numbers and triangle icons (without reducing gap between line numbers and text, if folding gutter disabled)

Usability: rename `"fadeFoldButtons"` pref to clearer `"hideUntilMouseover"`

And tweak code style issues:
- Remove most of ignored, unneeded package.json
- Preserve original CodeMirror copyright on CM-derived files
- Rename a few things for clarity
- Remove some unneeded code
- Tweak formatting & spelling in comments

**Update - adds a few more fixes/cleanups:**

More rigorous enable/disable flow for code folding: instead of piecemeal checking whether each bit has already initialized, atomically track whether init is done yet or not. Init/deinit all extant editors at once instead of doing each editor as it is activated, since splitview means editors may be already visible but not yet active. Fix listener leaks from repeated inits.

And keep menu items in a more predictable order:
- Add menu items at startup before user extensions load (at htmlReady instead of appReady), so core menu items like these don't move around as user extension load order changes
- Place QuickView menu item more explicitly, so it doesn't move unpredictably relative to Code Folding menu items (within the extensions/default folder, load order is arbitrary)
- Properly clean up menu divider, which used to get duplicated on each init pass

peterflynn commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -83,13 +76,12 @@ define(function (require, exports, module) {
         if (!editor || !saveFolds) { return; }
 
         var cm = editor._codeMirror;
-        if (!cm) {return; }
```
`editor._codeMirror` can never be null for any Editor, so I removed this check (the `editor` arg itself could be null -- e.g. in rare case where the `"documentRefreshed"` listener fires on a Document with no `_masterEditor` -- so I left that check above in place.

thehogfather replied:
good to know :)

peterflynn commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -232,27 +224,30 @@ define(function (require, exports, module) {
         var lnIndex = gutters.indexOf("CodeMirror-linenumbers");
-        gutters.splice(lnIndex + 1, 0, gutterName);
+        gutters.splice(lnIndex + 1, 0, GUTTER_NAME);
+        $(editor.getRootElement()).addClass("folding-enabled");
```
This enables us to apply different padding to line numbers when the folding gutter is present vs. not.

peterflynn commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -284,15 +279,18 @@ define(function (require, exports, module) {
+     * Initialise the extension
+     */
     function init() {
-        if (CodeMirror.fold.combine || !prefs.getSetting("enabled")) {
```
Per discussion in the PR, it seems like `CodeMirror.fold.combine` should be needed in the extension-registry copy of this, but not in the core version since it's guaranteed to always load first.

peterflynn commented:
A few other things we should probably do before this story closes down:

* Update wiki docs on writing themes (themes now need to consider changing the gutter triangle & inline placeholder colors)
* Add folding-related preferences to table in "How to Use Brackets" wiki (once the churn has settled)
* Add notes to extension-writing wiki docs about how to add folding support for new languages

I'm happy to help with those three.

Lastly, @abose are you already tracking a task for creating unit tests to go with this feature?  If not, should we create a bug or Trello card to track that work?

thehogfather commented:
@peterflynn Happy with these changes. I'll update PR #10850 later this evening and you can merge this.

thehogfather commented:
@peterflynn just to let you know I updated PR #10850. Merge your changes whenever you're ready.

MarcelGerber commented:
@peterflynn You got point 2 wrong for dark themes. Collapsed ones are _brighter_ (= _more_ outstanding) than expanded ones.

peterflynn commented:
@MarcelGerber That's by design -- collapsed should be more prominent.  So in the default light theme collapsed = darker, while in the dark theme collapsed = lighter.  Third-party themes can override these styles though.  I'll reword the bullet at top for clarity.

MarcelGerber commented:
Oh sorry, I got those wrong. What I meant was:
__Expanded__ ones are brighter (= more outstanding) than __collapsed__ ones.

peterflynn committed:
Improve visual appearance of code folding:
- Fainter, smaller gutter triangle icons
- Gutter triangle icons get less faint on mouseover (or when collapsed)
- Vertically align gutter triangle icons with text better
- Change inline collapsed placeholder to more readable arrow icon
- Better colors in dark theme
- Less horizontal gap between line numbers and triangle icons (without
  reducing gap between line numbers and text, if folding gutter disabled)

Usability: rename "fadeFoldButtons" pref to clearer "hideUntilMouseover"

And tweak code style issues:
- Remove most of ignored, unneeded package.json
- Preserve original CodeMirror copyright on CM-derived files
- Rename a few things for clarity
- Remove some unneeded code
- Tweak formatting & spelling in comments

peterflynn commented:
@MarcelGerber Hmm, that doesn't repro for me:

![screen shot 2015-04-09 at 5 31 49 pm](https://cloud.githubusercontent.com/assets/1197510/7079853/6eb46254-dede-11e4-95cd-5f8a27314896.png)

Do you have a third-party theme installed that might be interfering with the styles, or anything?

peterflynn committed:
More rigorous enable/disable flow for code folding: instead of piecemeal
checking whether each bit has already initialized, atomically track whether
init is done yet or not. Init/deinit all extant editors at once instead of
doing each editor as it is activated, since splitview means editors may be
already visible but not yet active. Fix listener leaks from repeated inits.

And keep menu items in a more predictable order:
- Add menu items at startup before user extensions load (at htmlReady instead
of appReady), so core menu items like these don't move around as user extension
load order changes
- Place QuickView menu item more explicitly, so it doesn't move unpredictably
relative to Code Folding menu items (within the extensions/default folder,
load order is arbitrary)
- Properly clean up menu divider, which used to get duplicated on each init
pass

peterflynn committed:
Fix stray tab chars
Remove unused `#code-folding-settings-dialog` CSS rule
Fix some comment whitespace

peterflynn commented:
@thehogfather Fyi, I pushed a few more updates to this PR -- see updated description at top for details and a link to the correct diff view.  It should now enable/disable dynamically much more cleanly, keeping the menus in proper order better and working better with splitview (and hopefully with slightly simpler code).

I've rebased on top of your PR now, so until that's merged, this PR here will show the diff from both combined.

abose commented:
![image](https://cloud.githubusercontent.com/assets/5336369/7087183/ffa7b50a-dfa5-11e4-80a9-31c6af6455da.png)
The icon is not rendered in my machine.

MarcelGerber commented:
@peterflynn Yeah you're right, it's an issue with my third-party theme (https://github.com/Brackets-Themes/Lion).
I'll file an issue over there as soon as this PR lands in master.

peterflynn commented:
@abose What OS are you on?  Does it show the other arrow icon (that the code on master uses) properly?

abose commented:
Windows 8.1. yes, it shows the arrow icon properly.

abose commented:
@peterflynn I was only fixing the unit test failures caused by the code folding extension in master.

I think it's best to add an issue to add unit test cases here as this is one of the core features now.

abose commented:
Now that PR #10850  has been merged, can this be merged?

MarcelGerber commented:
@abose Fixing the "code folding state doesn't get saved" issue (https://github.com/adobe/brackets/pull/10850#discussion_r28110161) is probably also something @peterflynn wants to fix in this PR (it's easy, even).

abose commented:
@peterflynn pull #10885 has the fix for the issue (#10850 (comment)), being a simple it can be done as part of this PR itself - or merge that pr after merging this .

peterflynn commented:
Let me just fix the Unicode issue @abose pointed out -- then I think this is good to merge.

peterflynn commented:
Ok, I spent a while playing with this and decided to switch from arrows to a simple "...":

![screen shot 2015-04-13 at 7 16 24 pm](https://cloud.githubusercontent.com/assets/1197510/7129212/f1849b00-e211-11e4-8bae-984e6317e8dd.png)

I actually like this a bit better than the wide arrows this PR had previously:

![screen shot 2015-04-13 at 7 17 02 pm](https://cloud.githubusercontent.com/assets/1197510/7129213/f184aaaa-e211-11e4-856d-0ad2e2790222.png)

And it's definitely way more legible than the original narrow arrows:

![screen shot 2015-04-13 at 7 18 17 pm](https://cloud.githubusercontent.com/assets/1197510/7129214/f18cd838-e211-11e4-886a-b70cf8bae144.png)

peterflynn committed:
Stop using supplemental Unicode arrow in the code-folding placeholder -
Windows may not have glyphs for it and simpler Unicode arrows don't look good.
Use simple "..." text instead (vertically aligned so it's more like an icon
than text).

abose commented:
Thanks @peterflynn for the quick fix.
+1 for 3 dots . 
Merging.
